### DISCUSSION TITLE
Where is the Bloom code located?

### DISCUSSION THREAD

**Discussion by Trainmaster9977** [ScanMountGoat/Smush-Material-Research#40] [Q&A]
Hello!

The bloom is currently getting in the way of my cell shading mod, and so I want to rewrite it. However, I don't know where the code _is_. But given that you have the code on the website, I'm guessing you would know?

Thanks!

**Comment by ScanMountGoat** [comment]
After the second color pass in RenderDoc, you should see a bunch of `glDrawElements(3)` calls. These calls perform the post processing and bloom. Bloom takes multiple steps, so there isn't a single shader. The bloom was reverse engineered from RenderDoc rather than the decompiled shaders themselves. It should be straightforward to find the shader file with matching code. If you want to disable bloom, you can also just ask users to use the _near render pass.

**Reply by Trainmaster9977** [reply]
The problem with _near is that it seems to just make the mesh in question completely transparent, which is obviously a bit of a problem. 

Also, time to dig through the shaders I guess. I have a few guesses for what might contain it, but guesses are guesses obv.

Thank you though!

**Reply by ScanMountGoat** [reply]
Is that the only render pass that does that? The problem with modifying bloom is that it affects all models. The near pass happens after bloom, so it essentially disables bloom for those meshes. You can also just keep the colors under the bloom threshold (roughly 75%). If you're experiencing a lot of bloom, the colors are probably too bright already. The threshold conveniently happens to be close to the color values that will display as white, so there shouldn't be too much unwanted blooming. You could rewrite the post processing shader that applies the LUT, but that will make the rest of the game too dark.

**Reply by Trainmaster9977** [reply]
It seems to only happen with near- and given it’s transparent I don’t think it’s just the lack of bloom.
